cmd/influxd/inspect/report_tsi/report_tsi.go

```diff
@@ -0,0 +1,451 @@
+	useBitmapN = 25
+)
+
+// reportTSI represents the program execution for "inspect report-tsi".
+type reportTSI struct {
+	// Flags
+	bucketId    string // required
+	enginePath  string
+	topN        int
+	concurrency int
+
+	// Variables for calculating and storing cardinalities
+	sfile         *tsdb.SeriesFile
+	shardPaths    map[uint64]string
+	shardIdxs     map[uint64]*tsi1.Index
+	cardinalities map[uint64]map[string]*cardinality
+}
+
+// NewReportTSICommand returns a new instance of Command with default setting applied.
+func NewReportTSICommand() *cobra.Command {
+	var arguments reportTSI
+	cmd := &cobra.Command{
+		Use:   "report-tsi",
+		Short: "Reports the cardinality of TSI files",
+		Long: `This command will analyze TSI files within a specified bucket, reporting the 
+cardinality of data within the files, segmented by shard and further by measurement.`,
+		Args: cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			arguments.shardPaths = map[uint64]string{}
+			arguments.shardIdxs = map[uint64]*tsi1.Index{}
+			arguments.cardinalities = map[uint64]map[string]*cardinality{}
+
+			return arguments.run()
+		},
+	}
+
+	cmd.Flags().StringVarP(&arguments.bucketId, "bucket-id", "b", "", "Required - specify which bucket to report on. A bucket id must be a base-16 string")
+	cmd.Flags().StringVar(&arguments.enginePath, "path", os.Getenv("HOME")+"/.influxdbv2/engine/data", "Path to engine.")
+	cmd.Flags().IntVarP(&arguments.topN, "top", "t", 0, "Limit results to top n")
+	cmd.Flags().IntVar(&arguments.concurrency, "c", runtime.GOMAXPROCS(0), "How many concurrent workers to run.")
+	cmd.MarkFlagRequired("bucket-id")
+
+	return cmd
+}
+
+// Run executes the command.
+func (report *reportTSI) run() error {
+	// Get all shards from specified bucket
+	dirEntries, err := os.ReadDir(filepath.Join(report.enginePath, report.bucketId, "autogen"))
+
+	if err != nil {
+		return err
+	}
+
+	for _, entry := range dirEntries {
+
+		if !entry.IsDir() {
+			continue
+		}
+
+		if entry.Name() == tsdb.SeriesFileDirectory || entry.Name() == "index" {
+			continue
+		}
+
+		id, err := strconv.Atoi(entry.Name())
+		if err != nil {
+			continue
+		}
+
+		report.shardPaths[uint64(id)] = filepath.Join(report.enginePath, report.bucketId, "autogen", entry.Name())
+	}
+
+	if len(report.shardPaths) == 0 {
+		fmt.Fprintf(os.Stderr, "No shards under %s\n", report.enginePath)
+		return nil
+	}
+
+	report.sfile = tsdb.NewSeriesFile(filepath.Join(report.enginePath, report.bucketId, tsdb.SeriesFileDirectory))
+
+	config := logger.NewConfig()
+	newLogger, err := config.New(os.Stderr)
+	if err != nil {
+		return err
+	}
+	report.sfile.Logger = newLogger
+
+	if err := report.sfile.Open(); err != nil {
+		return err
+	}
+	defer report.sfile.Close()
+
+	// Blocks until all work done.
+	report.calculateCardinalities(report.cardinalityByMeasurement)
+
+	// Print summary.
+	if err := report.printSummaryByMeasurement(); err != nil {
+		return err
```

Comment:
The indexes won't be closed in this case.

What I'd recommend is:
1. Move the declaration of `allIDs` above the call to `printSummaryByMeasurement`, remove the `Close()` calls from within the loop body
2. Change this from the short-hand way to a plain `err := report.printSummaryByMeasurement()`
3. Loop over `allIDs` before you check `err`, closing the indexes
4. Check `err != nil` and return